### PR TITLE
[IMP] registry: RO cursor fall back to RW

### PR DESCRIPTION
**Pull request by nim-odoo**
When using Odoo with a read-only replica, it is necessary for the replica to be online, otherwise read-only requests will fail.

This commit implements a simple failover mechanism: if a RO cursor cannot be acquired, we fall back to the RW cursor. We also deactivate the RO feature by setting `_db_readonly` to `None`. It will have the effect of deactivating the RO feature for the subsequent requests until Odoo is restarted or the worker is recycled.

In the context of the Postgres replication, such behavior is useful since synchronous replication doesn't guarantee that the replica is actually up-to-date in some edge cases. The most common case being the unexpected restart of the replica: there is a short period of time during which the replica accepts incoming requests but is still catching up with the master. Delaying the use of the replica should help reducing the race conditions in such cases.

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by kmagusiak** [comment]
I think that if there are issues with the RO replica, then it should not be set in the configuration.

The commit handles connection failures only to then always fallback to the RW replica.
If you remove the local variable `connection` you can simplify the code.

**Comment by Xavier-Do** [comment]
LGTM but:
- We may consider fixing it here 
https://github.com/odoo/odoo/blob/584a172274caefb23e5d91b609fc893160535416/odoo/http.py#L1793-L1795

Managing the OperationalError a similar way to the ReadOnlySqlTransaction

This would keep all retry mechanism in the same place. But in this case since it is when opening the cursor it is not a real problem to do it registry.py, it may avoid some overhead.

-  I'm not a big fan of disabling the _db_readonly. It's maybe not a big deal to spam this warning while the replica is not reachable instead of disabling the use of a replica for a while. But I'll leave it to your appreciation.

**Comment by Xavier-Do** [comment]
> db_replica_failover=none|request|permanent or something like that?

We usually try avoid adding command line argument, for such a specific use case it looks a little overkill. otoh having a way to disable this behavior could be interesting. 

We could also put the warning inside a dedicated logger so that it is possible to mute it.

File: odoo/modules/registry.py
Comment (Julien00859) [inline odoo/modules/registry.py:967]: Let's use the opportunity to document that `cursor(readonly=True)` doesn't matter when there is no replica.
```suggestion
          
            :param readonly: Attempt to acquire a cursor on a replica database. Acquire a read/write cursor on the primary database in case no replica exists or that no readonly cursor could be acquired.
```

File: odoo/modules/registry.py
Comment (Julien00859) [inline odoo/modules/registry.py:985]: I understand the motives for minimal diff, but since this is critical code I would rather we have the simplest code possible
```py
if readonly and self._db_readonly is not None:
    try:
        return self._db_readonly.cursor()
    except psycopg2.OperationalError:
        self._db_readonly = None
        _logger.warning(...)
return self._db.cursor()
```

**Review by Julien00859 — COMMENTED** [review]
How long is the replica instable? Is it a matter of milliseconds, seconds, minutes? Because the current strategy: to disable connecting to the replica until the worker is recycled (65k requests per default config, no limit on threads), seems fine only if the replica cannot be used for an extended time (>10 seconds, or whatever is the request timeout).

I see that `retrying` (`odoo/service/model.py`) already automatically retries on `OperationFailure` with an exponential back-off, but only for a limited subset of errors (`PG_CONCURRENCY_EXCEPTIONS_TO_RETRY`). Can we try and add this exception to that list? With back-pressure in mind, if we're talking <3sec of downtime, I think it would be better.

But in the end, you people at infra/internal know better! I don't want to block this PR if that's something you need. I'm fine experimenting this idea.

**Comment by nim-odoo** [comment]
> How long is the replica instable?

That is the point: we can't know in advance. It can go from a few seconds if the server has just a hiccup to more than 1 hour if there is a hardware error. If the server is down for an extended duration before restarting, there will be a time window during which the replica accepts connections while it is not yet up-to-date. In practice, it is possible to query the replica and get outdated results even if the replica is configured as 'synchronous'. It is due to the fact that the replica is agnostic its replication state (sync / async). The state is only known by the master.

`self._db_readonly = None` is intended to mitigate the risk since we delay the RO usage. If we consider the risk of getting outdated results acceptable, it can be removed.

**Comment by nim-odoo** [comment]
@odony 
To extend my point of view with a very analytic risk / reward perspective.

I start with the following hypothesis. At the moment:
- we, at Odoo, are the only ones using this feature.
- a RO replica is not required **yet** for our workload.

We can consider 2 types of unexpected issues:
1. recurrent short connection problems with the RO (< 1min, e.g. network connectivity issue, service crash...)
2. exceptional long downtime (> 30min, e.g. hardware error)

At the moment, it is likely that none of these issues will be recurrent. They might occur less than once a year, but they still need to be taken care of (because bad things always happen on a week-end). A failover RO -> RW is necessary for our (and particularly my) peace of mind.

**Case A**: we do not use `self._db_readonly = None`

If issue 1 arises, it allows querying the RO as fast as possible with a limited risk of getting outdated data.
In issue 2 arises, it allows querying the RO as fast as possible with a risk of getting outdated data increasing with the unavailability duration.

**Case B**: we use `self._db_readonly = None`

If issue 1 arises, it will take some time before we start querying the RO replica again. If the issue is recurrent, we might under use it.
If issue 2 arises, it will take some time before we start querying the RO replica again, but we have a limited risk of getting outdated data (the risk still exists though).

**Conclusion**

Considering the 2 hypothesis, case B is less risky than case A => I choose B.
In the worst case, we do not use the RO replica as much as we could, but since we do not need it (yet), that doesn't cause any problem. It gives us some time to collect data on the failover frequency. We will monitor these failovers anyway.
I might change my stance depending on the data we collect and the evolution of our workload in the future. As a wise CTO once said: it is software, we can always change it.

I have no plan on harshly defend this opinion. If we want to go for case A, fine by me.

**Comment by Xavier-Do** [comment]
The argument that if the replica is not available, it may became available with outdated data is a good point 

But I would add a last element to take in consideration: Execution consistency.

The fact that the cursor is readonly may impact the execution of some conditional piece of code. Lets imagine that it leads to a case where a readonly route fails with a readonly cursor. (or introduces invalid data)

A worker may continue to live for a while with a different behavior than the other ones, leading to random errors when reaching a specific worker.

It guess that it is unlikely enough to ignore this case.

But I think that to solve this we could replace the db_readonly by a readonly connection on the main.

`self._db_readonly = odoo.sql_db.db_connect(db_name, readonly=True)`

Again, not sure it is what we want but it would make this corner case a little less different that contacting the replica.
The downside is that if the requests needs to write, it will need to retry the transaction, but it shouldn't be a frequent case.

**Comment by Julien00859** [comment]
> Wouldn't a config parameter solve this dilemma, and allow us to introduce new failover policies later if we see the need to, without disrupting production systems?

@odony we cannot use a ICP since we have no connection to query the database, the configuration / environment variable is the only way I think. I think it would be a good idea to print a warning upon startup to say that the replica is an experimental feature.

**Comment by Xavier-Do** [comment]
> @odony we cannot use a ICP since we have no connection to query the database, the configuration / environment variable is the only way I think.

I had the same confusion but i think he is speaking about a command line param not an icp.

**Review by odony — APPROVED** [review]
alright, since we're starting from a situation where there's no degradation at all, I guess we can go with the current patch as a first step. It implements a semi-permanent failover to the RW connection, whose duration entirely depends on the lifetime of the workers (based on request_limit, cpu_limit, memory_limit, ...). This means it will be setup- and workload-dependent.

A good practice for admins will be to monitor the occurrence of the new warning "Failed to open a readonly cursor, falling back to read-write cursor" in the logs, and manually verify the state of their replicas asap.

Once we have more visibility on usage patterns (maybe in 2 months, 6 months, ...) we may need to implement additional failover policies. At that point it will still be time to add a new configuration option, with the current PR as the default.

PS: still need to be ~~rebased~~ squashed :)

**Review by Julien00859 — APPROVED** [review]
Thanks everyone for the discussion :)

@robodoo r+

**Pull request closed without merging**
